src/SharpLearning.Backend.Cntk.Test/CntkLayers.cs

```diff
@@ -0,0 +1,210 @@
+        /// <param name="shape">Shape (`int` or `tuple` of `ints`): vector or tensor dimension of the output of this layer</param>
+        /// <param name="activation">Activation (:class:`~cntk.ops.functions.Function`, defaults to identity): optional function to apply at the end, e.g. `relu`</param>
+        /// <param name="init">init (scalar or NumPy array or :mod:`cntk.initializer`, defaults to :func:`~cntk.initializer.glorot_uniform` ): initial value of weights `W`</param>
+        /// <param name="inputRank">input_rank (int, defaults to `None`): number of inferred axes to add to W (`map_rank` must not be given)</param>
+        /// <param name="mapRank"> map_rank (int, defaults to `None`): expand W to leave exactly `map_rank` axes (`input_rank` must not be given)</param>
+        /// <param name="bias">bias (bool, optional, defaults to `True`): the layer will have no bias if `False` is passed here</param>
+        /// <param name="initBias">init_bias (scalar or NumPy array or :mod:`cntk.initializer`, defaults to 0): initial value of weights `b`</param>
+        /// <returns>cntk.ops.functions.Function: A function that accepts one argument and applies the operation to it</returns>
+        public Function Dense(Variable x, int shape, Func<Variable, Function> activation = null, CNTKDictionary init = null,
+            int inputRank = 0, int mapRank = 0, bool bias = true, CNTKDictionary initBias = null)
+        {
+            if(inputRank != 0 && mapRank != 0)
+            {
+                throw new ArgumentException("Dense: inputRank and mapRank cannot be specified at the same time.");
+            }
+
+            var outputShape = NDShape.CreateNDShape(new int[] { shape });
+            var outputRank = outputShape.Dimensions.Count;
+
+            var inputRanks = (inputRank != 0) ? inputRank : 1;
+            var dimensions = Enumerable.Range(0, inputRanks).Select(v => NDShape.InferredDimension).ToArray(); // infer all dimensions.
+            var inputShape = NDShape.CreateNDShape(dimensions);
+            
+            int inferInputRankToMap;
+
+            if (inputRank != 0)
+                inferInputRankToMap = -1; // means map_rank is not specified; input_rank rules.
+            else if (mapRank == 0)
+                inferInputRankToMap = 0;  // neither given: default to 'infer W to use all input dims'.
+            else
+                inferInputRankToMap = mapRank;  // infer W to use all input dims except the first static 'map_rank' ones.
+
+            var wDimensions = outputShape.Dimensions.ToList();
+            wDimensions.AddRange(inputShape.Dimensions);
+            var wShape = NDShape.CreateNDShape(wDimensions);
+
+            init = init ?? m_getDefaultInitializer(); // should also be possible to provide a fixed value.
+            var w = new Parameter(wShape, m_dataType, init, m_device, "w");
+
+            // Weights and input is in reversed order compared to the original python code.
```

Comment:
It seems that the python API is reversing all dimensions internally, and this is the reason for the reverse order of dimensions when comparing to the C# or the C++ API. The reason this is done in the CNTK python interface was to be similar to other deep learning toolkits. More information can be found here [input shape dimensions different between python and csharp](https://github.com/Microsoft/CNTK/issues/2790)

From the issue/link:
**"The input it isn't different as it might seem. CNTK internally is column major, so (224,224,3) column major is the same as (3,224,224) row major. In python API, we reverse all the shapes in order to be similar to other DNN frameworks. The C# binding doesn't do that and is similar to the C++ API"**

Reply:
@nietras An important detail regarding the Python vs. the C# CNTK API.